glib/src/main_context_futures.rs

```diff
@@ -230,19 +241,108 @@ impl TaskSource {
 
                 // This will panic if the future was a local future and is called from
                 // a different thread than where it was created.
-                Pin::new(&mut self.future).poll(&mut context)
+                let res = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+                    Pin::new(&mut self.future).poll(&mut context)
+                }));
+                match res {
+                    Ok(Poll::Ready(res)) => {
+                        let _ = self.return_tx.take().unwrap().send(Ok(res));
+                        Poll::Ready(())
+                    }
+                    Ok(Poll::Pending) => Poll::Pending,
+                    Err(e) => {
+                        let _ = self.return_tx.take().unwrap().send(Err(e));
+                        Poll::Ready(())
+                    }
+                }
             })
             .expect("current thread is not owner of the main context")
     }
 }
 
+// rustdoc-stripper-ignore-next
+/// A handle to a task running on a [`MainContext`].
+///
+/// Like [`std::thread::JoinHandle`] for a task rather than a thread. The return value from the
+/// task can be retrieved by awaiting on this object. Dropping the handle "detaches" the task,
+/// allowing it to complete but discarding the return value.
+#[derive(Debug)]
+pub struct JoinHandle<T> {
+    rx: oneshot::Receiver<std::thread::Result<T>>,
+    id: Cell<Option<NonZeroU32>>,
+}
+
+impl<T> JoinHandle<T> {
+    #[inline]
+    fn new(
+        ctx: &MainContext,
+        (source, rx): (Source, oneshot::Receiver<std::thread::Result<T>>),
+    ) -> Self {
+        let id = source.attach(Some(ctx));
+        let id = Cell::new(Some(unsafe { NonZeroU32::new_unchecked(id.as_raw()) }));
+        Self { rx, id }
+    }
+    // rustdoc-stripper-ignore-next
+    /// Returns the internal source ID.
+    ///
+    /// Returns `None` if the handle was aborted.
+    #[inline]
+    pub unsafe fn as_raw_source_id(&self) -> Option<u32> {
```

Comment:
Would it be useful to also have a getter for it as `SourceId`?

Reply:
Not a getter, but there could be `into_source_id`. `SourceId` avoids implementing `Clone` and `Copy` so the source cannot accidentally be removed twice

Reply:
Makes sense